File: Tests/AINotebookCoreTests/NotebookStoreProvidersTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class NotebookStoreProvidersTests: XCTestCase {

    private func makeStore() throws -> NotebookStore {
        try NotebookStore(path: .inMemory)
    }

    func testSaveAndFetchRoundTrips() throws {
        let store = try makeStore()
        let cfg = ProviderConfig(
            type: .openwebui, name: "LAN server", baseURL: "http://192.168.1.50:3000"
        )
        try store.saveProvider(cfg)
        let loaded = try XCTUnwrap(try store.provider(id: cfg.id))
        XCTAssertEqual(loaded.type, .openwebui)
        XCTAssertEqual(loaded.name, "LAN server")
        XCTAssertEqual(loaded.baseURL, "http://192.168.1.50:3000")
        XCTAssertTrue(loaded.enabled)
        XCTAssertFalse(loaded.privacyAcknowledged)
    }

    func testProvidersListsSeedPlusSavedOrderedByCreation() throws {
        let store = try makeStore()
        let cfg = ProviderConfig(type: .anthropic, name: "Claude", baseURL: "https://api.anthropic.com")
        try store.saveProvider(cfg)
        let all = try store.providers()
        XCTAssertEqual(all.count, 2)
        XCTAssertEqual(all.first?.id, ProviderConfig.ollamaId)
        XCTAssertEqual(all.last?.id, cfg.id)
    }

    func testUpdatePreservesPrivacyAcknowledgement() throws {
        let store = try makeStore()
        var cfg = ProviderConfig(type: .openai, name: "OpenAI", baseURL: "https://api.openai.com")
        try store.saveProvider(cfg)
        try store.acknowledgePrivacy(providerId: cfg.id)
        cfg.name = "OpenAI renamed"
        try store.saveProvider(cfg)   // cfg still carries privacyAcknowledged == false
        let loaded = try XCTUnwrap(try store.provider(id: cfg.id))
        XCTAssertEqual(loaded.name, "OpenAI renamed")
        XCTAssertTrue(loaded.privacyAcknowledged, "edit must not reset the consent flag")
    }

    func testDeleteRefusesBuiltInOllama() throws {
        let store = try makeStore()
        XCTAssertThrowsError(try store.deleteProvider(id: ProviderConfig.ollamaId))
        XCTAssertNotNil(try store.provider(id: ProviderConfig.ollamaId))
    }

    func testDeleteRemovesRow() throws {
        let store = try makeStore()
        let cfg = ProviderConfig(type: .openwebui, name: "X", baseURL: "http://h:3000")
        try store.saveProvider(cfg)
        try store.deleteProvider(id: cfg.id)
        XCTAssertNil(try store.provider(id: cfg.id))
    }

    func testUnknownTypeStringLoadsAsOpenAICompatible() throws {
        let store = try makeStore()
        try store.runOnDatabase { db in
            try db.execute(
                sql: """
                INSERT INTO providers(id, type, name, base_url, enabled, privacy_acknowledged, created_at)
                VALUES ('future-id', 'grok', 'Future', 'https://x', 1, 0, datetime('now'))
                """
            )
        }
        let loaded = try XCTUnwrap(try store.provider(id: "future-id"))
        XCTAssertEqual(loaded.type, .openaiCompatible)
    }
}

File: windows/tests/AINotebook.Core.Tests/Storage/NotebookStoreProvidersTests.cs
using AINotebook.Core.Providers;
using AINotebook.Core.Storage;
using Xunit;

namespace AINotebook.Core.Tests.Storage;

public class NotebookStoreProvidersTests
{
    private static NotebookStore Fresh() => new(StorePath.InMemory);

    private static ProviderConfig MakeConfig(ProviderType type = ProviderType.OpenAI, string name = "OpenAI") =>
        new(Guid.NewGuid().ToString(), type, name, "https://api.openai.com",
            true, false, DateTime.UtcNow);

    [Fact]
    public void SaveAndFetchRoundTrips()
    {
        using var store = Fresh();
        var cfg = MakeConfig(ProviderType.OpenWebUI, "LAN server") with { BaseUrl = "http://192.168.1.50:3000" };
        store.SaveProvider(cfg);
        var loaded = store.Provider(cfg.Id)!;
        Assert.Equal(ProviderType.OpenWebUI, loaded.Type);
        Assert.Equal("LAN server", loaded.Name);
        Assert.Equal("http://192.168.1.50:3000", loaded.BaseUrl);
        Assert.True(loaded.Enabled);
        Assert.False(loaded.PrivacyAcknowledged);
    }

    // NotebookStoreProvidersTests.testUpdatePreservesPrivacyAcknowledgement (macOS parity)
    [Fact]
    public void UpdatePreservesPrivacyAcknowledgement()
    {
        using var store = Fresh();
        var cfg = MakeConfig(ProviderType.OpenAI, "OpenAI");
        store.SaveProvider(cfg);
        store.AcknowledgePrivacy(cfg.Id);

        // Re-save (e.g. a rename) still carries PrivacyAcknowledged == false,
        // exactly as AddProviderViewModel.SaveConfirmedAsync always constructs
        // cloud-type configs — the store must not let this clobber the flag.
        var renamed = cfg with { Name = "OpenAI renamed" };
        store.SaveProvider(renamed);

        var loaded = store.Provider(cfg.Id)!;
        Assert.Equal("OpenAI renamed", loaded.Name);
        Assert.True(loaded.PrivacyAcknowledged, "edit must not reset the consent flag");
    }

    [Fact]
    public void DeleteRemovesRow()
    {
        using var store = Fresh();
        var cfg = MakeConfig(ProviderType.OpenWebUI, "X") with { BaseUrl = "http://h:3000" };
        store.SaveProvider(cfg);
        store.DeleteProvider(cfg.Id);
        Assert.Null(store.Provider(cfg.Id));
    }

    // Regression: the v11 migration's data step used to seed the built-in
    // Ollama provider's created_at with DateTime.UtcNow.ToString("yyyy-MM-dd
    // HH:mm:ss") — no milliseconds — while SqliteDate.FromDb does a strict
    // ParseExact on "yyyy-MM-dd HH:mm:ss.fff". Every fresh DB (in-memory or
    // on-disk) therefore carried an unreadable seeded row, and Provider()/
    // Providers() threw FormatException the moment they touched it. This
    // must succeed on a brand-new store with no other setup.
    [Fact]
    public void SeededOllamaProvider_IsReadableRightAfterMigration()
    {
        using var store = Fresh();

        var cfg = store.Provider(ProviderConfig.OllamaId);

        Assert.NotNull(cfg);
        Assert.Contains(store.Providers(), p => p.Id == ProviderConfig.OllamaId);
    }
}
